src/main/java/educationalproject/programmingstuff/servicies/OrderServiceImpl.java

```diff
@@ -0,0 +1,96 @@
+    private OrderResponseMapper orderResponseMapper;
+
+    @Override
+    @Transactional
+    public OrderResponseDto book(OrderCreateRequestDto newOrder) {
+        List<RequestedItemDto> requested = newOrder.getRequestedItemDtos();
+        List<CommodityItem> available = getAvailableCommodityItems(requested);
+        return writeOff(requested, available);
+    }
+
+    private List<CommodityItem> getAvailableCommodityItems(List<RequestedItemDto> requestedItemDtos) {
+        Set<Long> itemsId = requestedItemDtos.stream()
+                .map(RequestedItemDto::getItemId)
+                .collect(Collectors.toSet());
+        return commodityItemRepository.getCommodityItemsByItemIdIn(itemsId);
+    }
+
+    private OrderResponseDto writeOff(List<RequestedItemDto> requested, List<CommodityItem> available) {
+        Map<Long, Integer> requestedItemsIdMappedToQuantity = makeMapOf(requested);
+        WriteOffOrThrowException(available, requestedItemsIdMappedToQuantity);
+
+        List<OrderItem> orderItems = available.stream()
+                .map(availableCommodityItem -> {
+                    Item item = availableCommodityItem.getItem();
+                    return OrderItem.builder()
+                            .item(item)
+                            .quantity(requestedItemsIdMappedToQuantity.get(item.getId()))
+                            .build();
+                })
+                .collect(Collectors.toList());
+
+        Order order = Order.builder().orderItems(orderItems).build();
+
+        orderRepository.save(order);
+
+        return orderResponseMapper.makeOrderResponseOf(order);
+    }
+
+    private Map<Long, Integer> makeMapOf(List<RequestedItemDto> commodityItems) {
+        return commodityItems.stream()
+                .collect(Collectors.toMap(
+                        RequestedItemDto::getItemId,
+                        RequestedItemDto::getQuantity));
+    }
+
+    private void WriteOffOrThrowException(
+            List<CommodityItem> available, Map<Long, Integer> requestedItemsIdMappedToQuantity) {
+        available.forEach(
+                availableItem -> {
+                    long itemId = availableItem.getItem().getId();
+                    int requestedQuantity = requestedItemsIdMappedToQuantity.get(itemId);
+                    int availableQuantity = availableItem.getQuantity();
+                    int remainingQuantity = availableQuantity - requestedQuantity;
+                    if (remainingQuantity < 0) {
+                        throw new NotEnoughItemsAtStoreException(
+                                "Requested " + requestedQuantity +
+                                        " items with item number " + itemId +
+                                        " but there is only " + availableQuantity +
+                                        " items at the storehouse.");
+                    }
+                    availableItem.setQuantity(remainingQuantity);
+                });
```

Comment:
Just asking... What if, for any reason, our user asked us to write off some items with non-existing ids?

Reply:
I will check this during test writing. For now, I assume two possibilities: whether Hibernate will return exception while perform getAvailableCommodityItems() or I should make a new one.